### PR TITLE
YARN-10522. Document for Flexible Auto Queue Creation in Capacity Scheduler.

### PR DESCRIPTION
**Pull request by brumi1024**
### For code changes:

- [ ] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 44s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  markdownlint  |   0m  0s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  32m 30s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 28s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  46m 15s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 14s |  |  the patch passed  |
| -1 :x: |  blanks  |   0m  0s | [/blanks-eol.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/1/artifact/out/blanks-eol.txt) |  The patch has 3 line(s) that end in blanks. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  mvnsite  |   0m 18s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  13m 24s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 33s |  |  The patch does not generate ASF License warnings.  |
|  |   |  62m 50s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3330 |
| Optional Tests | dupname asflicense mvnsite codespell markdownlint |
| uname | Linux 8a8cac681c87 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / d67a03d73b95bbcf6a870a91ccd1d7ef657f3ed8 |
| Max. process+thread count | 547 (vs. ulimit of 5500) |
| modules | C: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site U: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/1/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 57s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  markdownlint  |   0m  0s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  31m 47s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 29s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  45m 49s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 13s |  |  the patch passed  |
| -1 :x: |  blanks  |   0m  0s | [/blanks-eol.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/2/artifact/out/blanks-eol.txt) |  The patch has 3 line(s) that end in blanks. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  mvnsite  |   0m 21s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  13m 37s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 36s |  |  The patch does not generate ASF License warnings.  |
|  |   |  62m 58s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3330 |
| Optional Tests | dupname asflicense mvnsite codespell markdownlint |
| uname | Linux f236535a8650 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / d5390530f83470ac120011c7fa7074df12e7e730 |
| Max. process+thread count | 720 (vs. ulimit of 5500) |
| modules | C: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site U: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/2/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m  9s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  markdownlint  |   0m  0s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  33m 20s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 30s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  49m 46s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 14s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 19s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  14m 51s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 33s |  |  The patch does not generate ASF License warnings.  |
|  |   |  68m 18s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3330 |
| Optional Tests | dupname asflicense mvnsite codespell markdownlint |
| uname | Linux 1fd267351106 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 97759a26a42eb445da128c16c1173b95ebb24b98 |
| Max. process+thread count | 611 (vs. ulimit of 5500) |
| modules | C: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site U: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/3/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/CapacityScheduler.md [hunk 1]
```diff
@@ -123,17 +123,25 @@ Configuration
...
 
-  * Resource Allocation using Absolute Resources configuration
+  * Configuring Resource Allocation
 
- `CapacityScheduler` supports configuration of absolute resources instead of providing Queue *capacity* in percentage. As mentioned in above configuration section for `yarn.scheduler.capacity.<queue-path>.capacity` and `yarn.scheduler.capacity.<queue-path>.max-capacity`, administrator could specify an absolute resource value like `[memory=10240,vcores=12]`. This is a valid configuration which indicates 10GB Memory and 12 VCores.
+  `CapacityScheduler` supports three different resource allocation configuration modes: percentage values (*relative mode*), weights and absolute resources.
+
+  Relative mode provides a way to describe queue's resources as a fraction of its parent's resources. For example if *capacity* is set as 50.0, users queue has 50% of its parent, root's resources set as minimum capacity.
```
Comment (9uapaw) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/CapacityScheduler.md]: The last sentence is hard to understand, could you rephrase it please?

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/CapacityScheduler.md [hunk 2]
```diff
@@ -681,6 +689,55 @@ list of current scheduling edit policies as a comma separated string in `yarn.re
...
+
+| Property | Description |
+|:---- |:---- |
+| `yarn.scheduler.capacity.<queue-path>.auto-queue-creation-v2.template.<queue-property>` | *Optional* parameter: Specifies a queue property (like capacity, maximum-capacity, user-limit-factor, maximum-am-resource-percent ...  - Refer **Queue Properties** section) inherited by the auto-created **parent** and **leaf** queues. |
+| `yarn.scheduler.capacity.<queue-path>.auto-queue-creation-v2.leaf-template.<queue-property>` | *Optional* parameter: Specifies a queue property inherited by auto-created **leaf** queues. |
+| `yarn.scheduler.capacity.<queue-path>.auto-queue-creation-v2.parent-template.<queue-property>` |  *Optional* parameter: Specifies a queue property inherited by auto-created **parent** queues. |
+
+Using the following example configuration snippet will instruct the `CapacityScheduler` to:
+* enable the flexible auto queue creation for root.parent
+* create **all** of the dynamic queues below root.parent with 80% as the maximum capacity, because of the wildcard queue path (root.parent.*)
```
Comment (9uapaw) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/CapacityScheduler.md]: I believe it does not stand true. This wildcard sets maximum-capacity for all descendants of every parent queue under parent.

Comment (brumi1024) [reply hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/CapacityScheduler.md]: This line is for the property yarn.scheduler.capacity.root.parent.*.auto-queue-creation-v2.template.maximum-capacity, which has a simple template tag, not the parent-template. Doesn't this apply to all of the queues, not just the parents?

Or do you mean that this overrides the static queues as well?

Comment (9uapaw) [reply hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/CapacityScheduler.md]: I mean root.parent.*.template.max-capacity will set the max-capacity for root.parent.parent-2.leaf. To set maximum-capacity for all queues under root.parent, simply set the template of root.parent. Templating could be confusing in some cases, because if you define a template property, it will set the properties one level lower (hence I advised to provide an example).

Comment (brumi1024) [reply hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/CapacityScheduler.md:714]: That's what I meant by all, as the two points below has directly under part in bold text. But you are right I'll clear it up a bit more as this could be misunderstood.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m  2s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  33m 49s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 24s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  50m 10s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 12s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  15m 55s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 30s |  |  The patch does not generate ASF License warnings.  |
|  |   |  69m 29s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3330 |
| Optional Tests | dupname asflicense mvnsite codespell markdownlint |
| uname | Linux f65c1cd52dee 4.15.0-147-generic #151-Ubuntu SMP Fri Jun 18 19:21:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / a5713d1d20537b5ec54eed2c8151faae7d3531e8 |
| Max. process+thread count | 575 (vs. ulimit of 5500) |
| modules | C: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site U: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/4/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Review by 9uapaw — APPROVED** [review]
Thanks for the revisions @brumi1024. It looks good to me +1.

**Comment by szilard-nemeth** [comment]
Thanks @brumi1024 for working on this.
Latest state looks good to me, committed to trunk.
Thanks @9uapaw for the review.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m  4s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  markdownlint  |   0m  0s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  37m 34s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 29s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  57m 13s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 15s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   0m 23s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  16m 51s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 36s |  |  The patch does not generate ASF License warnings.  |
|  |   |  77m 57s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3330 |
| Optional Tests | dupname asflicense mvnsite codespell markdownlint |
| uname | Linux 4fe6b43e4632 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 19f3d16e77f0a6b513ea164da909b77347648d3f |
| Max. process+thread count | 750 (vs. ulimit of 5500) |
| modules | C: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site U: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3330/5/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.
